cornac/models/trirank/recom_trirank.py

```diff
@@ -0,0 +1,326 @@
+            col.append(j)
+            data.append(matrix[i, j] / (row_norm[i] * col_norm[j]))
+
+        return csr_matrix((data, (row, col)), shape=matrix.shape)
+
+    def _create_matrices(self, train_set):
+        from time import time
+
+        start_time = time()
+        if self.verbose:
+            print("Building matrices started!")
+        sentiment_modality = train_set.sentiment
+        n_users = train_set.num_users
+        n_items = train_set.num_items
+        n_aspects = sentiment_modality.num_aspects
+
+        X_row = []
+        X_col = []
+        X_data = []
+        Y_row = []
+        Y_col = []
+        Y_data = []
+        for uid, isid in tqdm(
+            sentiment_modality.user_sentiment.items(),
+            disable=not self.verbose,
+            desc="Building matrices",
+        ):
+            for iid, sid in isid.items():
+                aos = sentiment_modality.sentiment[sid]
+                aids = set(aid for aid, _, _ in aos)  # Only one per review/sid
+                for aid in aids:
+                    X_row.append(iid)
+                    X_col.append(aid)
+                    X_data.append(1)
+                    Y_row.append(uid)
+                    Y_col.append(aid)
+                    Y_data.append(1)
+
+        # Algorithm 1: Offline training line 2
+        X = csr_matrix((X_data, (X_row, X_col)), shape=(n_items, n_aspects))
+        Y = csr_matrix((Y_data, (Y_row, Y_col)), shape=(n_users, n_aspects))
+
+        # Algorithm 1: Offline training line 3
+        X.data = np.log2(X.data) + 1
+        Y.data = np.log2(Y.data) + 1
+
+        # Algorithm 1: Offline training line 4
+        if self.verbose:
+            print("Building symmetric normalized matrices R, X, Y")
+        self.R = self._symmetrical_normalization(train_set.csr_matrix)
+        self.X = self._symmetrical_normalization(X)
+        self.Y = self._symmetrical_normalization(Y)
+
+        if self.verbose:
+            total_time = time() - start_time
+            print("Building matrices completed in %d s" % total_time)
+
+    def fit(self, train_set, val_set=None):
+        """Fit the model to observations.
+
+        Parameters
+        ----------
+        train_set: :obj:`cornac.data.Dataset`, required
+            User-Item preference data as well as additional modalities.
+
+        val_set: :obj:`cornac.data.Dataset`, optional, default: None
+            User-Item preference data for model selection purposes (e.g., early stopping).
+
+        Returns
+        -------
+        self : object
+        """
+        Recommender.fit(self, train_set, val_set)
+        self._init()
+
+        if not self.trainable:
+            return self
+
+        # Offline training: Build item-aspect matrix X and user-aspect matrix Y
+        self._create_matrices(train_set)
+        return self
+
+    def _online_recommendation(self, user):
+        # Algorithm 1: Online recommendation line 5
+        p_0 = self.train_set.csr_matrix[user]
```

Comment:
We might need a bracket for `user`, i.e., `p_0 = self.train_set.csr_matrix[[user]]` for 1D slicing. Please check if it works or not.

Reply:
There is no 1D slicing for this case as it will be converted to a 1D array later `p_0 = p_0.toarray().squeeze()`